security/pkg/stsservice/tokenmanagers/google/tokenmanager.go

```diff
@@ -0,0 +1,303 @@
+	federatedTokenEndpoint = "https://securetoken.googleapis.com/v1/identitybindingtoken"
+	// https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
+	accessTokenEndpoint = "https://iamcredentials.googleapis.com/v1/projects/-/" +
+		"serviceAccounts/service-%s@gcp-sa-meshdataplane.iam.gserviceaccount.com:generateAccessToken"
+)
+
+// TokenManager supports token exchange with Google OAuth 2.0 authorization server.
+type TokenManager struct {
+	hTTPClient  *http.Client
+	trustDomain string
+	// tokens is the cache for fetched tokens.
+	// map key is timestamp of token, map value is tokenInfo.
+	tokens           sync.Map
+	gCPProjectNumber string
+}
+
+// CreateTokenManager creates a token manager that fetches token from a Google OAuth 2.0 authorization server.
+func CreateTokenManager(trustDomain string, gCPProjectNumber string) (*TokenManager, error) {
+	caCertPool, err := x509.SystemCertPool()
+	if err != nil {
+		tokenManagerLog.Errorf("Failed to get SystemCertPool: %v", err)
+		return nil, err
+	}
+	tm := &TokenManager{
+		hTTPClient: &http.Client{
+			Timeout: httpTimeOutInSec * time.Second,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs: caCertPool,
+				},
+			},
+		},
+		trustDomain:      trustDomain,
+		gCPProjectNumber: gCPProjectNumber,
+	}
+	return tm, nil
+}
+
+type federatedTokenResponse struct {
+	AccessToken     string `json:"access_token"`
+	IssuedTokenType string `json:"issued_token_type"`
+	TokenType       string `json:"token_type"`
+	ExpiresIn       int64  `json:"expires_in"` // Expiration time in seconds
+}
+
+// GenerateToken takes STS request parameters and fetches token, returns StsResponseParameters in JSON.
+func (tm *TokenManager) GenerateToken(parameters stsservice.StsRequestParameters) ([]byte, error) {
+	tokenManagerLog.Debugf("Start to fetch token with STS request parameters: %v", parameters)
+	ftResp, err := tm.fetchFederatedToken(parameters)
+	if err != nil {
+		return nil, err
+	}
+	atResp, err := tm.fetchAccessToken(ftResp)
+	if err != nil {
+		return nil, err
+	}
+	return tm.generateSTSResp(atResp)
+}
+
+// constructFederatedTokenRequest returns an HTTP request for federated token.
+// Example of a federated token request:
+// POST https://securetoken.googleapis.com/v1/identitybindingtoken
+// Content-Type: application/json
+// {
+//    audience: <trust domain>
+//    grantType: urn:ietf:params:oauth:grant-type:token-exchange
+//    requestedTokenType: urn:ietf:params:oauth:token-type:access_token
+//    subjectTokenType: urn:ietf:params:oauth:token-type:jwt
+//    subjectToken: <jwt token>
+//    Scope: https://www.googleapis.com/auth/cloud-platform
+// }
+func (tm *TokenManager) constructFederatedTokenRequest(parameters stsservice.StsRequestParameters) *http.Request {
+	reqScope := scope
+	if len(parameters.Scope) != 0 {
+		reqScope = parameters.Scope
+	}
+	query := map[string]string{
+		"audience":           tm.trustDomain,
```

Comment:
What is this tm.trustDomain? Is it in this format? identityNamespaces:$Identity-namespace-id:$issuer-url?

Reply:
This is passed from env variable, the format could be identityNamespaces:$Identity-namespace-id:$issuer-url